curvefs/src/metaserver/dumpfile.cpp

```diff
@@ -0,0 +1,501 @@
+    }
+
+    LOG(INFO) << "Save success";
+    return DUMPFILE_ERROR::OK;
+}
+
+DUMPFILE_ERROR DumpFile::WaitSaveDone(pid_t childpid) {
+    int status;
+    if (waitpid(childpid, &status, WUNTRACED) == -1) {
+        LOG(ERROR) << "[child] Unexpected waitpid() failure";
+        return DUMPFILE_ERROR::WAITPID_FAILED;
+    }
+
+    if (WIFEXITED(status)) {
+        LOG(INFO) << "[child] Save process (pid=" << childpid
+                  << ") exit, exitCode = " << WEXITSTATUS(status);
+        return WEXITSTATUS(status) == 0 ? DUMPFILE_ERROR::OK :
+                                          DUMPFILE_ERROR::FAILED;
+    }
+
+    int signal = -1;
+    if (WIFSIGNALED(status)) {
+        signal = WTERMSIG(status);
+    } else if (WIFSTOPPED(status)) {
+        signal = WSTOPSIG(status);
+        if (kill(childpid, SIGKILL) == -1) {
+            LOG(ERROR) << "kill (pid=" << childpid << ") failed";
+        }
+    }
+
+    LOG(ERROR) << "[child] Save process (pid=" << childpid
+               << ") exited by receive signal, signal = " << signal;
+
+    return DUMPFILE_ERROR::UNEXPECTED_SIGNAL;
+}
+
+void DumpFile::SignalHandler(int signo, siginfo_t* siginfo, void* ucontext) {
+    auto pid = (siginfo && siginfo->si_pid) ? siginfo->si_pid : -1;
+    LOG(INFO) << "Signal " << signal << " received from " << pid;
+    exit(2);
+}
+
+DUMPFILE_ERROR DumpFile::InitSignals() {
+    // The BRPC has install the SIGINT and SIGTERM, so we should
+    // catch them and to avoid release resources repeatedly in child process
+    std::vector<::curve::common::Signal> signals {
+        { SIGINT, SignalHandler },
+        { SIGTERM, SignalHandler },
+    };
+
+    if (!::curve::common::Process::InitSignals(signals)) {
+        LOG(ERROR) << "Init signals failed";
+        return DUMPFILE_ERROR::FAILED;
+    }
+
+    return DUMPFILE_ERROR::OK;
+}
+
+DUMPFILE_ERROR DumpFile::CloseSockets() {
+    std::vector<std::string> names;
+    pid_t pid = getpid();
+    if (fs_->List("/proc/self/fd", &names) != 0) {
+        return DUMPFILE_ERROR::LIST_FAILED;
+    }
+
+    auto nFail = 0;
+    struct stat info;
+    for (const auto& name : names) {
+        int fd = std::stoi(name);
+        if (fs_->Fstat(fd, &info) != 0) {
+            nFail++;
+        } else if (S_ISSOCK(info.st_mode) && fs_->Close(fd) != 0) {
+            return DUMPFILE_ERROR::CLOSE_FAILED;
+        }
+    }
+
+    // There is one small pitfall to what out for:
+    // the /proc/self/fd directory itself also is
+    // an open file descriptor (but fstast() will fail)
+    // while scanning it
+    return nFail > 1 ? DUMPFILE_ERROR::FSTAT_FAILED :
+                       DUMPFILE_ERROR::OK;
+}
+
+void DumpFile::SaveWorker(std::shared_ptr<Iterator> iter) {
+    auto retCode = InitSignals();
+    if (retCode != DUMPFILE_ERROR::OK) {
+        exit(1);
+    }
+
+    // We should close all socket fds in child process
+    // to ensure free connection success
+    retCode = CloseSockets();
```

Comment:
> Does parent and child process both can receive rpc request and process request simultaneously between `fork` and `CloseSockets` ?
> If so, data(inode/dentry) will be inconsistent between parent process and its child.

No, the child process will can't receive RPC request，the reason for close socket is once we fork, the reference count of the socket fd will plus one, so when parent process invoke the close, it will not free the connection actually.